rfcs/2022-04-20-12187-log-namespacing.md

```diff
@@ -0,0 +1,442 @@
+The key is still very important information that will be used by the sink.
+
+In general, if the data would be used by a sink when writing the data, it is considered "metadata" (stored on the event).
+Otherwise, it iss "vector metadata" (stored on the event metadata).
+
+#### Vector Metadata
+
+This is anything that is not really part of the log, but describes something useful _about_ the log, such as where
+it came from, or when it was received. These will be stored in event metadata. Some improvements will be needed to
+event metadata to make this easier to use, as described below.
+
+### Codecs
+
+#### Bytes
+
+With the "Legacy" namespace, this codec decodes the input as-is, and places it under a "message" key (configurable with global log schema).
+With the "Vector" namespace, this will be placed under the "data" key. If there is no additional metadata stored
+on the event, then it will be placed on the root instead. The placement will be decided for each source independently.
+
+#### Json
+
+With the "Legacy" namespace, this codec decodes the input as-is, and places it under a "message" key (configurable with global log schema).
+With the "Vector" namespace, this will be placed under the "data" key. If there is no additional metadata stored
+on the event, then it will be placed on the root instead. The placement will be decided for each source independently.
+
+#### Syslog
+
+All data will be placed at the root. Since the structure is known, it's easy to prevent naming collisions here.
+Syslog has a "message" field which might be useful to apply an additional codec to, but this is currently not supported,
+and no support is being added here.
+
+#### Native / Native JSON
+
+When these are used in the "Vector" source, the behavior will not change, and events will be passed through as-is.
+For any other source, deserialized data should be nested under "data", unless it can be elided (same rules as the JSON source).
+
+### Event Metadata
+
+There is currently minimal support for event metadata. Several changes will need to be made.
+
+Changes needed immediately:
+
+- Support arbitrarily nested values
+- Functions that access metadata (`get_metadata_field`, `remove_metadata_field`, and `set_metadata_field`) should support full paths as keys, and return
+  the `any` type instead of just `string. This is technically a breaking change.
+
+With these changes, using metadata can still be a bit annoying since the returned type will always be `any`, even if the
+value is set and read in the same VRL program.
+
+Future changes (out of scope for now):
+
+- Add a special path syntax to reference metadata instead of using functions.
+- Expand schema support to metadata
+- Expand semantic meaning to metadata
+
+
+### Implementation
+
+A proof of concept for the Datadog Agent Logs source is being worked on alongside this RFC: [12218](https://github.com/vectordotdev/vector/pull/12218)
+
+
+### Examples
```

Comment:
I would prefer if these examples include the original event as received by Vector, to make it clear what is being done to get to the desired state.

Reply:
With the latest update, the transformation is fairly simple, and I tried to add comments for anything not very straightforward. If you have any specific questions for an example, I'd be glad to clarify but I think adding all of the inputs will make this very messy.